gtsam/nonlinear/NonlinearFactor.h

```diff
@@ -272,503 +273,439 @@ class GTSAM_EXPORT NoiseModelFactor: public NonlinearFactor {
 
 
 /* ************************************************************************* */
-
 /**
- * A convenient base class for creating your own NoiseModelFactor with 1
- * variable.  To derive from this class, implement evaluateError().
+ * A convenient base class for creating your own NoiseModelFactor
+ * with n variables.  To derive from this class, implement evaluateError().
+ *
+ * For example, a 2-way factor could be implemented like so:
+ *
+ * ~~~~~~~~~~~~~~~~~~~~{.cpp}
+ * class MyFactor : public NoiseModelFactorN<double, double> {
+ *  public:
+ *   using Base = NoiseModelFactorN<double, double>;
+ *
+ *   MyFactor(Key key1, Key key2, const SharedNoiseModel& noiseModel)
+ *       : Base(noiseModel, key1, key2) {}
  *
- * Templated on a values structure type. The values structures are typically
- * more general than just vectors, e.g., Rot3 or Pose3,
- * which are objects in non-linear manifolds (Lie groups).
+ *   Vector evaluateError(
+ *       const double& x1, const double& x2,
+ *       boost::optional<Matrix&> H1 = boost::none,
+ *       boost::optional<Matrix&> H2 = boost::none) const override {
+ *     if (H1) *H1 = (Matrix(1, 1) << 1.0).finished();
+ *     if (H2) *H2 = (Matrix(1, 1) << 2.0).finished();
+ *     return (Vector(1) << x1 + 2 * x2).finished();
+ *   }
+ * };
+ * ~~~~~~~~~~~~~~~~~~~~
+ *
+ * These factors are templated on a values structure type. The values structures
+ * are typically more general than just vectors, e.g., Rot3 or Pose3, which are
+ * objects in non-linear manifolds (Lie groups).
  */
-template<class VALUE>
-class NoiseModelFactor1: public NoiseModelFactor {
-
-public:
-
-  // typedefs for value types pulled from keys
-  typedef VALUE X;
-
-protected:
-
-  typedef NoiseModelFactor Base;
-  typedef NoiseModelFactor1<VALUE> This;
-
-public:
+template <class... VALUES>
+class NoiseModelFactorN : public NoiseModelFactor {
+ public:
+  /// N is the number of variables (N-way factor)
+  enum { N = sizeof...(VALUES) };
+
+  /** The type of the i'th template param can be obtained as X<I> */
+  template <int I, typename std::enable_if<(I < N), bool>::type = true>
+  using X = typename std::tuple_element<I, std::tuple<VALUES...>>::type;
+
+ protected:
+  using Base = NoiseModelFactor;
+  using This = NoiseModelFactorN<VALUES...>;
+
+  /* Like std::void_t, except produces `boost::optional<Matrix&>` instead. Used
+   * to expand fixed-type parameter-packs with same length as VALUES */
+  template <typename T>
+  using optional_matrix_type = boost::optional<Matrix&>;
+
+  /* Like std::void_t, except produces `Key` instead. Used to expand fixed-type
+   * parameter-packs with same length as VALUES */
+  template <typename T>
+  using key_type = Key;
+
+ public:
   /// @name Constructors
   /// @{
 
-  /** Default constructor for I/O only */
-  NoiseModelFactor1() {}
-
-  ~NoiseModelFactor1() override {}
+  /**
+   * Default Constructor for I/O
```

Comment:
This can be a one line docstring
```
/// Default Constructor for I/O
```

Reply:
I've generally seen that we use `///` for single line comments. We definitely don't follow the Google Style in this case. 😅

Reply:
done